backend/src/cache/storage/execution_cache_store.go

```diff
@@ -35,8 +35,8 @@ type ExecutionCacheStore struct {
 	time util.TimeInterface
 }
 
-func (s *ExecutionCacheStore) GetExecutionCache(executionCacheKey string) (*model.ExecutionCache, error) {
-	r, err := s.db.Table("execution_caches").Where("ExecutionCacheKey = ?", executionCacheKey).Rows()
+func (s *ExecutionCacheStore) GetExecutionCache(executionCacheKey string, maxCacheStaleness int64) (*model.ExecutionCache, error) {
+	r, err := s.db.Table("execution_caches").Where("ExecutionCacheKey = ? AND MaxCacheStaleness = ?", executionCacheKey, maxCacheStaleness).Rows()
```

Comment:
I think this is not be the full logic. But you can fix that in a next PR.
The full logic is:
all_cache_entries = all entries with same ExecutionCacheKey
viable_cache_entries = entries from all_cache_entries where maxCacheStaleness == -1 || s.time.Now().UTC().Unix() - startedAtInSec <= maxCacheStaleness

try to return latest viable entry with same max_cache_staleness
else try to copy and return latest viable entry
else execute pod

Reply:
A question here: what if for a single step user previously set the max_cache_staleness=-1 and later on changes mind and set max_cache_staleness=0 to disable cache? If we always search all_cache_entries, once a step has unexpired entry it will be always using cache.

Reply:
>If we always search all_cache_entries, once a step has unexpired entry it will be always using cache.

I'm not sure that's the case. If user specifies maxCacheStaleness=0 we never return a cache entry.
We only ever return _viable_ cache entries - the ones where `maxCacheStaleness == -1 || entry.startedAtInSec >= s.time.Now().UTC().Unix() - maxCacheStaleness`. Notice that `maxCacheStaleness` here comes from the pod creation request, not from cache entry.
So if `maxCacheStaleness == 0` the predicate simplifies to `entry.startedAtInSec >= s.time.Now().UTC().Unix()` which should not be the case for any entry.

P.S. We can also explicitly special-case `maxCacheStaleness == 0` so that we completely skip the DB lookup in that case.